pkg/controller/machineset/machineset_controller.go

```diff
@@ -207,48 +296,353 @@ func (c *MachineSetController) handleErr(err error, key interface{}) {
+		utilruntime.HandleError(fmt.Errorf("invalid machineset key %q: either namespace or name is missing", key))
+		return nil
 	}
 
 	machineSet, err := c.machineSetsLister.MachineSets(ns).Get(name)
 	if errors.IsNotFound(err) {
-		glog.V(2).Infof("Node group %v has been deleted", key)
+		logger.Debugln("machine set has been deleted")
+		c.jobControl.ObserveOwnerDeletion(key)
 		return nil
 	}
 	if err != nil {
 		return err
 	}
 
-	// Deep-copy otherwise we are mutating our cache.
-	// TODO: Deep-copy only when needed.
-	ng := machineSet.DeepCopy()
+	shouldProvisionMachineSet := c.shouldProvision(machineSet)
+
+	jobFactory, err := c.getJobFactory(machineSet)
+	if err != nil {
+		return err
+	}
+
+	job, isJobNew, err := c.jobControl.ControlJobs(key, machineSet, shouldProvisionMachineSet, jobFactory)
+	if err != nil {
+		return err
+	}
+
+	if !shouldProvisionMachineSet {
+		return nil
+	}
+
+	switch {
+	// New job has not been created, so an old job must exist. Set the cluster
+	// to not provisioning yet as the old job is deleted.
+	case job == nil:
+		return c.setMachineSetToNotProvisioning(machineSet)
+	// Job was not newly created, so sync machine set status with job.
+	case !isJobNew:
+		logger.Debugln("provisioning job exists, will sync with job")
+		return c.syncMachineSetStatusWithJob(machineSet, job)
+	// MachineSet should have a job to provision the current spec but it was not
+	// found.
+	case isMachineSetProvisioning(machineSet):
+		return c.setJobNotFoundStatus(machineSet)
+	// New job created for new provisioning
+	default:
+		return nil
+	}
+}
+
+func (c *MachineSetController) clusterForMachineSet(machineSet *clusteroperator.MachineSet) (*clusteroperator.Cluster, error) {
+	controllerRef := metav1.GetControllerOf(machineSet)
+	if controllerRef.Kind != clusterKind.Kind {
+		return nil, nil
+	}
+	cluster, err := c.clustersLister.Clusters(machineSet.Namespace).Get(controllerRef.Name)
+	if err != nil {
+		return nil, err
+	}
+	if cluster.UID != controllerRef.UID {
+		// The controller we found with this Name is not the same one that the
+		// ControllerRef points to.
+		return nil, nil
+	}
+	return cluster, nil
+}
+
+func (c *MachineSetController) shouldProvision(machineSet *clusteroperator.MachineSet) bool {
+	if machineSet.Status.ProvisionedJobGeneration == machineSet.Generation {
+		return false
+	}
+	cluster, err := c.clusterForMachineSet(machineSet)
+	if err != nil {
+		return false
+	}
+	if !cluster.Status.Provisioned || cluster.Status.ProvisionedJobGeneration != cluster.Generation {
+		return false
+	}
+	switch machineSet.Spec.NodeType {
+	case clusteroperator.NodeTypeMaster:
+		return true
+	case clusteroperator.NodeTypeCompute:
+		masterMachineSet, err := c.machineSetsLister.MachineSets(machineSet.Namespace).Get(cluster.Status.MasterMachineSetName)
+		if err != nil {
+			return false
+		}
+		return masterMachineSet.Status.Installed && masterMachineSet.Status.InstalledJobGeneration == masterMachineSet.Generation
```

Comment:
Should the == be a >=? Or does an increase in generation imply re-installation and thus ==? (could use a comment either way)

Reply:
It wouldn't make sense for the InstalledJobGeneration to be greater than the Generation. That would mean that the master controller set a InstalledJobGeneration to a Generation that has not yet existed.

I've added a comment.

Reply:
Sorry I should have put <=, I was wondering if the machine set gen could be further ahead than installed gen, but it looks like we can probably assume this will trigger when they're equal and after that we don't need to do anything.

Reply:
This check is there *specifically* to prevent provisioning the compute nodes in the interregnum between when the master nodes were installed with the old spec and when they are installed with the new spec. The `masterMachineSet.Status.Installed` flag is invalidated by the spec change. But the machine set controller may try to sync the compute node before the master controller gets around to syncing the updated master machine set to clear the flag.